Fill.py

```diff
@@ -63,26 +63,174 @@ def fill_restrictive(multiworld: MultiWorld, base_state: CollectionState, locati
+                batched_placements_remaining = sum(items_per_player_in_batch.values())
+
+            # If there was an existing batch sweep state, it must be re-created because it will have collected items
+            # that are being placed in the current batch. The placement loop is responsible for creating the sweep state
+            # because the sweep state may get destroyed by a swap.
+            batch_sweep_state = None
+
         if one_item_per_player:
             # grab one item per player
             items_to_place = [items.pop()
                               for items in reachable_items.values() if items]
         else:
-            next_player = multiworld.random.choice([player for player, items in reachable_items.items() if items])
-            items_to_place = []
-            if item_pool:
-                items_to_place.append(reachable_items[next_player].pop())
-
+            # grab the next item
+            if next_player_override is None:
+                next_player = multiworld.random.choice([player for player, items in reachable_items.items() if items])
+                player_remaining_items_in_batch = items_per_player_in_batch[next_player]
+
+                if player_remaining_items_in_batch == 0:
+                    # This player still has items to place, but their items in the batch have been exhausted.
+                    # A new batch needs to be built. This prevents unfairness in picked items at the boundary between
+                    # two batches.
+                    batched_placements_remaining = 0
+                    next_player_override = next_player
+                    continue
+            else:
+                # We are starting a new batch because we tried to pick an item that was not in the previous batch.
+                next_player = next_player_override
+                # Clear the override.
+                next_player_override = None
+                player_remaining_items_in_batch = items_per_player_in_batch[next_player]
+                assert player_remaining_items_in_batch > 0
+
+            items_to_place = [reachable_items[next_player].pop()]
+            items_per_player_in_batch[next_player] = player_remaining_items_in_batch - 1
         for item in items_to_place:
+            for p, batch_pool_item in enumerate(batch_item_pool):
+                if batch_pool_item is item:
+                    batch_item_pool.pop(p)
+                    break
             for p, pool_item in enumerate(item_pool):
                 if pool_item is item:
                     item_pool.pop(p)
                     break
 
+        batched_placements_remaining -= 1
+
+        explore_locations = multiworld.get_filled_locations(item.player) if single_player_placement else None
+
+        # Create the initial batch sweep state or recreate it if it was destroyed by a swap.
+        if batch_sweep_state is None:
```

Comment:
```suggestion
        if batch_sweep_state is None:
            assert batch_base_state is not None, "should be set when creating the batch"
```
If someone changes the code to break this, this will make the error message more clear.
This cleans up the type checking so we can see other problems more easily.
(and `assert` statements are removed in optimization)

Reply:
Actually, can this be moved outside of the `if` block?
There's another place at line 342 where it's assumed to be set.
Putting it outside of this `if` block would cover both of those.

Reply:
Un-placing an item that `batch_sweep_state` has already collected puts the un-placed item back into `batch_item_pool` if there is space in the current batch. (There is space in the current batch if the player the item belongs to did not have enough items remaining in `reachable_items[unplaced_item.player]` to fill out the full number-of-items-per-player of the current batch.)

The issue is that the next `maximum_exploration_state` is created by sweeping from `batch_sweep_state` with the items in `batch_item_pool + unplaced_items` collected into its inventory. This would lead to the un-placed item being in both `batch_sweep_state`'s inventory and being in `batch_item_pool` simultaneously, so that one un-placed item would be collected twice by the next `maximum_exploration_state` for the current batch.

To prevent this, `batch_sweep_state` is destroyed when swapping (`batch_sweep_state = None`), which is why the `if batch_sweep_state is None:` check is necessary.

---

I think I can safely move the `batch_sweep_state = None` into the conditional branch that puts un-placed items back into the current batch, so that `batch_sweep_state` is not destroyed when the un-placed item has to be placed in a later batch.

---

I don't think I can use `batch_sweep_state.remove()` to remove the un-placed item because that will not remove any other items that could only be collected because the un-placed item had been collected.

I think if `batch_sweep_state.remove()` was used, `maximum_exploration_state` would still end up correct to start with, despite sweeping from the invalid `batch_sweep_state`, because `maximum_exploration_state` collects the un-placed item that was put into `batch_item_pool`, but when it is time to re-place the un-placed item, the un-placed item is removed from `batch_item_pool` prior to creating `maximum_exploration_state`, so `maximum_exploration_state` would be incorrect when placing the item.

`batch_sweep_state` is also the preferred state to sweep from to create `swap_state` when swapping. Allowing `batch_sweep_state` to become invalid would prevent this swap optimisation.

Reply:
I haven't been able to wrap my head around all of that yet, but the question I asked (unclearly) is more about `batch_base_state` rather than `batch_sweep_state`
(You didn't mention `batch_base_state` in your reply.)

`batch_base_state` is initialized to `None`, but it looks like it will always be a `CollectionState` (not `None`) before it ever gets referenced.
(And once it changes to `CollectionState`, it never changes to `None` again.)
That's all good if that's the case, and I was kind of asking you to confirm that.

The issue that came to mind is that it's slightly fragile with the initialization to a `CollectionState` being inside an `if` statement (line 115).
With AP code always changing (especially `Fill.py`), I can imagine things getting moved or tweaked by someone in the future so that initialization gets missed.

So outside of that initializing `if`, just before it gets referenced, I'm suggesting an `assert` to make sure `batch_base_state is not None` before it gets referenced.

At first, I only noticed the reference inside this `if` block
```python
        if batch_sweep_state is None:
            # `locations=None` defaults to `multiworld.get_filled_locations()`, so only get it once for both the batch
            # sweep state and the maximum exploration state.
            if explore_locations is None:
                explore_locations = multiworld.get_filled_locations()
            batch_sweep_state = sweep_from_pool(batch_base_state, locations=explore_locations)
```
where it would be a problem if `batch_base_state` were `None`

Because I only first noticed it there, I suggested the `assert` inside that `if` block.
But then I noticed references to `batch_base_state` further down, so the `assert` right before the `if` statement would be before all of the references.

```python
        assert batch_base_state is not None, "should be set when creating the batch"

        # Create the initial batch sweep state or recreate it if it was destroyed by a swap.
        if batch_sweep_state is None:
		    ...
```

Reply:
Ah, I failed to properly read what the suggested change was doing.

`batch_base_state` should always be non-None after the first batch has been created.

I needed to assign the `batch_base_state` outside the loop because the assignment that occurs when creating the first batch occurs inside an if statement. Using the `batch_base_state` variable outside that if statement would have PyCharm complaining that `batch_base_state` is potentially being referenced before assignment.

I assigned to `None` rather than a `CollectionState` instance to reduce the likelihood of silent errors occurring and generation crashing further on.

I can add an assertion with a message.